controller_connections.go

```diff
@@ -90,19 +92,116 @@ func (c *controller) handleIncoming(con net.Conn) {
 		return
 	}
 
-	peer := newPeer(c.net, c.peerStatus, c.peerData)
-	// we are never expecting a reject-alternate for incoming connections
-	if _, err := peer.StartWithHandshake(ep, con, true); err != nil {
-		c.logger.WithError(err).Debugf("Handshake failed for address %s, stopping", ep)
-		peer.Stop()
+	// don't bother with alternatives for incoming connections
+	peer, _, err := c.handshake(host, con, true)
+
+	if err != nil {
+		c.logger.WithError(err).Debugf("inbound connection from %s failed handshake", host)
+		con.Close()
 		return
 	}
 
 	c.logger.Debugf("Incoming handshake success for peer %s, version %s", peer.Hash, peer.prot.Version())
+}
+
+func (c *controller) handshake(ip string, con net.Conn, incoming bool) (*Peer, []Endpoint, error) {
+	tmplogger := c.logger.WithField("addr", ip)
+	timeout := time.Now().Add(c.net.conf.HandshakeTimeout)
+
+	nonce := make([]byte, 8) // loopback detection
+	binary.LittleEndian.PutUint64(nonce, c.net.instanceID)
+
+	// upgrade connection to a metrics connection
+	metrics := NewMetricsReadWriter(con)
+
+	handshake := newHandshake(c.net.conf, nonce)
+	decoder := gob.NewDecoder(metrics) // pipe gob through the metrics writer
+	encoder := gob.NewEncoder(metrics)
+	con.SetWriteDeadline(timeout)
+	con.SetReadDeadline(timeout)
+
+	failfunc := func(err error) (*Peer, []Endpoint, error) {
+		tmplogger.WithError(err).Debug("Handshake failed")
+		con.Close()
+		return nil, nil, err
+	}
+
+	err := encoder.Encode(handshake)
+	if err != nil {
+		return failfunc(fmt.Errorf("Failed to send handshake to incoming connection"))
+	}
+
+	var reply Handshake
+	err = decoder.Decode(&reply)
```

Comment:
From your answer on the issue I created you said:
```
Both sides will send their handshake first, then read the first incoming parcel.
```
What are the guarantees about that first incoming parcel? What type of parcel will it be? It seems that the parcel will be kind of dropped because it won't be processed by the routing logic, is that a problem? Are we guaranteed what at least 1 initial parcel will be sent in time before timeout?

Reply:
> What are the guarantees about that following first incoming parcel? What type of parcel will it be?

P2P1 is hardcoded to send a peer request in the "go online" function:
https://github.com/FactomProject/factomd/blob/6f0164ba51dc1776d1ca44a3555ae1dd510bc5fb/p2p/connection.go#L389-L391


> It seems that the parcel will be "dropped" (read, parsed but not really processed) because it won't be processed by the routing logic, is that a problem?

It's a minor issue for V9 nodes. At the moment, P2P1 only has a global timer where it asks for a peer request on establishing a connection, and also every 3 minutes for *all* connections at once:

https://github.com/FactomProject/factomd/blob/6f0164ba51dc1776d1ca44a3555ae1dd510bc5fb/p2p/controller.go#L604-L610

So there is a scenario where a V9 node connects to an all V10 network and doesn't get an immediate response. The result is that it's only connected to seed servers and waits up to 3 minutes before asking the seed servers to share their peers.

It doesn't affect existing nodes but it's not a good place to be in. I'm going to address this issue and fix it, so that P2P2 will respond correctly to V9 "handshakes" with a peer response.

Reply:
this is now in: 
https://github.com/WhoSoup/factom-p2p/blob/4393dfa181676216d93a6ee38fc8feb54520398c/controller_connections.go#L197-L202